### PR TITLE
Provider: added end session support

### PR DESCRIPTION
**Pull request by dzen-it**
This is a popular feature that is useful when working with the library.

### CODE REVIEW

File: oidc.go [hunk 1]
```diff
@@ -238,6 +242,73 @@ func (p *Provider) UserInfo(ctx context.Context, tokenSource oauth2.TokenSource)
...
 
+// EndSessionOptions is optional parameters for the Initiated Logout by https://openid.net/specs/openid-connect-session-1_0.html#RPLogout
+type EndSessionOptions struct {
+	IDTokenHint string
+	RedirectURL string
+	State       string
+}
+
+// EndSession uses the token source to query the provider's end session endpoint.
+func (p *Provider) EndSession(ctx context.Context, tokenSource oauth2.TokenSource, options EndSessionOptions) error {
```
Comment (ericchiang) [inline oidc.go:253]: nit: rename "options" to "o"

File: oidc.go [hunk 2]
```diff
@@ -238,6 +242,73 @@ func (p *Provider) UserInfo(ctx context.Context, tokenSource oauth2.TokenSource)
...
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	return nil
+}
+
+func makeEndSessionEndpoint(endSessionURL string, options EndSessionOptions) (endURL string, err error) {
```
Comment (ericchiang) [inline oidc.go:287]: nit: rename "options" to "o"
nit: rename "makeEndSessionEndpoint" to "endSessionEndpoint"

Also please use shorter variable names

```go
	if endSessionURL == "" {
		return "", errors.New("oidc: end session endpoint is not supported by this provider")
	}

	u, err := url.Parse(endSessionURL)
	if err != nil {
		return "", fmt.Errorf("oidc: wrong format of end session endpoint %s: %v", p.endSessionURL, err)
	}
	q := u.Query()
	if o.IDTokenHint != "" {
		q.Set("id_token_hint", o.IDTokenHint)
	}
	if o.RedirectURL != "" {
		q.Set("post_logout_redirect_uri", o.RedirectURL)
	}
	if o.State != "" {
		q.Set("state", o.State)
	}
	u.RawQuery = q.Encode()
	return u.String(), nil
```

Comment (ericchiang) [reply oidc.go:287]: Also, please add a test for this method.

File: oidc.go [hunk 3]
```diff
@@ -238,6 +242,73 @@ func (p *Provider) UserInfo(ctx context.Context, tokenSource oauth2.TokenSource)
...
+	}
+
+	token, err := tokenSource.Token()
+	if err != nil {
+		return fmt.Errorf("oidc: get access token: %v", err)
+	}
+
+	token.SetAuthHeader(req)
+
+	resp, err := doRequest(ctx, req)
```
Comment (ericchiang) [inline oidc.go:271]: I'm not sure it's correct for the library to make this request. The spec specifically calls out that this should be initiated by the browser, not the server

```
An RP can notify the OP that the End-User has logged out of the site, and might want to log out of the OP as well. In this case, the RP, after having logged the End-User out of the RP, redirects the End-User's User Agent to the OP's logout endpoint URL. This URL is normally obtained via the end_session_endpoint element of the OP's Discovery response, or may be learned via other mechanisms.
```
https://openid.net/specs/openid-connect-session-1_0-17.html#RPLogout

**Comment by ericchiang** [comment]
Also when you mention "This is a popular feature that is useful when working with the library." do you have a list of providers that actually implement this feature? :)

**Comment by dzen-it** [comment]
I looked carefully at the implementation of the top providers, and  you are right, the "end session" should be cooked on the client side. Thank you for your time.

**Pull request closed without merging**